### PR TITLE
Add ECS types

### LINKED ISSUE #1349: Add ECS Resources
### What resource do you need?

ECS resources (specifically Cluster, Service, and TaskDefinition resources)

### What is your use case?

Running containerized apps outside of Kubernetes. The terrajet provider features ECS resources, but they do not have resource selector references and wouldn't work for our use case anyway, since we already have things like XRDs defined for ALBs in the native provider.

### Would you be willing to contribute it using [code generator](https://github.com/crossplane/provider-aws/blob/master/CODE_GENERATION.md)?
Yes I plan on generating the code and putting up a PR for this.

### PR DESCRIPTION
**Pull request by EdgeJ**
This commit adds the Cluster, Service, and TaskDefinition resources from
the ECS family.

Signed-off-by: EdgeJ <5093048+EdgeJ@users.noreply.github.com>

### Description of your changes

Fixes #1349 

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Run `make reviewable test` to ensure this PR is ready for review.

### How has this code been tested

Tested using example yaml files added in this commit to create and destroy all resources.
[contribution process]: https://git.io/fj2m9

### CODE REVIEW

**Comment by EdgeJ** [comment]
Not sure why the `check-diff` task is failing. Locally, I can run it with no problems:
```
➜ make check-diff
13:10:54 [ .. ] installing goimports
13:11:04 [ OK ] installing goimports
Cloning into 'code-generator'...
remote: Enumerating objects: 4003, done.
remote: Counting objects: 100% (22/22), done.
remote: Compressing objects: 100% (17/17), done.
<SNIP>
13:15:14 [ OK ] go mod tidy
13:15:14 [ .. ] checking that branch is clean
13:15:14 [ OK ] branch is clean
```

On the other hand, I get unrelated errors trying to run `make test`:
```
➜ make test
13:25:57 [ .. ] verify dependencies have expected content
all modules verified
13:26:00 [ OK ] go modules dependencies verified
13:26:01 [ .. ] installing go-junit-report
13:26:02 [ OK ] installing go-junit-report
13:26:02 [ .. ] installing gocover-cobertura
13:26:03 [ OK ] installing gocover-cobertura
13:26:03 [ .. ] go test unit-tests
package github.com/crossplane-contrib/provider-aws/pkg/clients/mock: build constraints exclude all Go files in /Users/jedge/git/crossplane/provider-aws/pkg/clients/mock
13:26:04 [FAIL]
make[1]: *** [go.test.unit] Error 1
make: *** [test] Error 2
```

**Comment by haarchri** [comment]
can you run make generate again ?

**Comment by EdgeJ** [comment]
> can you run make generate again ?

@haarchri I've tried running `make generate`, `make check-diff`, and every combination of make's I can think of. I've also tried locally, in a docker container, and an ec2 instance with a clean checkout of the repo on my branch. I can't manage to recreate the CI failures anywhere. Without more output from the CI job to understand what the unclean diff is, I'm spinning my wheels at this point, unable to recreate the problem.

**Comment by haarchri** [comment]
In ci Output you can See that the following Files modified 

apis/ecs/v1alpha1/zz_generated.deepcopy.go
 M package/crds/ecs.aws.crossplane.io_clusters.yaml
 M package/crds/ecs.aws.crossplane.io_services.yaml
 M package/crds/ecs.aws.crossplane.io_taskdefinitions.yaml

Which GO Version you using ?

**Comment by EdgeJ** [comment]
Local Go version:
```
➜ go version
go version go1.17 darwin/amd64
```

Running on Docker via golang:1.17-buster:
```
root@aa29417540cd:/go/src/provider-aws# go version
go version go1.17.11 linux/amd64
```

**Comment by haarchri** [comment]
okay same on my side - let me re-run the check-diff in ci

```
19:38:28 [ .. ] verify dependencies have expected content
all modules verified
19:38:32 [ OK ] go modules dependencies verified
19:38:32 [ .. ] go generate darwin_amd64
19:38:53 [ OK ] go generate darwin_amd64
19:38:53 [ .. ] go mod tidy
19:38:54 [ OK ] go mod tidy
19:38:54 [ .. ] checking that branch is clean
19:38:54 [ OK ] branch is clean
```

File: apis/ecs/v1alpha1/custom_types.go
Comment (haarchri) [inline apis/ecs/v1alpha1/custom_types.go:32]: think we can add Ref/Selector here ? https://github.com/crossplane-contrib/provider-aws/blob/master/apis/elbv2/v1alpha1

File: apis/ecs/v1alpha1/custom_types.go
Comment (haarchri) [inline apis/ecs/v1alpha1/custom_types.go:149]: think we can use here Ref/Selector ?

File: apis/ecs/v1alpha1/custom_types.go
Comment (EdgeJ) [reply apis/ecs/v1alpha1/custom_types.go:32]: Sure. I'll add this.

File: apis/ecs/v1alpha1/custom_types.go
Comment (EdgeJ) [reply apis/ecs/v1alpha1/custom_types.go:149]: I think IAM is an example of the inconsistent AWS APIs, from the api documentation (https://docs.aws.amazon.com/sdk-for-go/api/service/ecs/), it looks like this is effectively a boolean flag, with an enum value of "ENABLED" or "DISABLED", rather than a pointer to an IAM resource:
```
const (
    // EFSAuthorizationConfigIAMEnabled is a EFSAuthorizationConfigIAM enum value
    EFSAuthorizationConfigIAMEnabled = "ENABLED"

    // EFSAuthorizationConfigIAMDisabled is a EFSAuthorizationConfigIAM enum value
    EFSAuthorizationConfigIAMDisabled = "DISABLED"
)
```

File: apis/ecs/v1alpha1/custom_types.go
Comment (EdgeJ) [reply apis/ecs/v1alpha1/custom_types.go]: Done.

**Comment by nabuskey** [comment]
I can reproduce this on Ubuntu 18.04. Steps:

```bash
git -c protocol.version=2 fetch --no-tags --prune --progress --no-recurse-submodules --depth=1 origin +ec0c388743e51239bd377bb3b1dc8b5f38951a8b:refs/remotes/pull/1368/merge
git checkout --progress --force refs/remotes/pull/1368/merge
make check-diff
...
18:35:45 [ .. ] checking that branch is clean
 M apis/ecs/v1alpha1/zz_generated.deepcopy.go
 M package/crds/ecs.aws.crossplane.io_clusters.yaml
 M package/crds/ecs.aws.crossplane.io_services.yaml
 M package/crds/ecs.aws.crossplane.io_taskdefinitions.yaml
```
You probably just need to rebase.

**Comment by nabuskey** [comment]
Diff is rather large but looks like it's missing the new policy fields.
[123.txt](https://github.com/crossplane-contrib/provider-aws/files/9022687/123.txt)

**Comment by haarchri** [comment]
yes so rebase & run make generate again should fix the problem

**Comment by EdgeJ** [comment]
ohhhhh. I hadn't thought of it performing diff checks against upstream code.  I'll try rebasing and pushing up to see if that fixes things. Meanwhile, I think it might make sense to hold this for now until I see what happens with #1376 since that might affect how the ecs resources reference loadbalancer resources.

**Review by haarchri — APPROVED** [review]
LGTM thanks for the contribution @EdgeJ

**Comment by haarchri** [comment]
@EdgeJ did you need to add something after merge https://github.com/crossplane-contrib/provider-aws/pull/1376 ? 
i tested the your branch here - with success

**Comment by EdgeJ** [comment]
I rebased on the master branch before #1376 was merged, but I haven't yet addressed the difference in types as the result of that change yet.

**Comment by EdgeJ** [comment]
Updated the `LoadBalancerName` func to use the updated `AtProvider` api type. @haarchri this should be GTG now.

**Pull request merged**
